`/wp/v2/pattern-directory/patterns` endpoint: slug parameter has no effect on the response

anton-vlasenko opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/55617

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

anton-vlasenko committed:
Fix a bug in the WP_REST_Pattern_Directory_Controller:get_items.
We must also implode slugs as implode doesn't work with multi-dimensional arrays.

anton-vlasenko committed:
Update ticket reference.

anton-vlasenko committed:
Fix code style.

anton-vlasenko committed:
Add `slug` to the collection parameters. Previously, it was missing.

anton-vlasenko committed:
Improve PHPDOC blocks.

anton-vlasenko committed:
Update wp-api-generated.js schema.

anton-vlasenko committed:
Update wp-api-generated.js schema.

anton-vlasenko committed:
Update wp-api-generated.js schema.

anton-vlasenko committed:
Update comment.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -337,6 +328,11 @@ public function get_collection_params() {
 
+		$query_params['slug'] = array(
+			'description' => __( 'Limit results to those matching a pattern (slug).' ),
+			'type'        => array( 'string', 'array' ),
```
We should make this just an `array`. The REST API will accept a string value and upgrade it to an array. That way the controller can always treat the `slug` parameter as an `array`.

anton-vlasenko replied:
Good to know. Thank you, @TimothyBJacobs.
Fixed in 5e923bc73da1322a2b64bfaf9fc090abfadfc57b

anton-vlasenko committed:
"Slug" must be of type array.
Strings will be automatically converted to an array.
Per suggestion here: https://github.com/WordPress/wordpress-develop/pull/2625#discussion_r858069357

anton-vlasenko committed:
Change the scope of the WP_REST_Pattern_Directory_Controller::get_transient_key method as it would be helpful to make it available to child classes.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -346,4 +342,40 @@ public function get_collection_params() {
+	protected function get_transient_key( $query_args ) {
+
+		if ( array_key_exists( 'slug', $query_args ) ) {
+			if ( ! is_array( $query_args['slug'] ) ) {
+				$query_args['slug'] = explode( ',', (string) $query_args['slug'] );
+			}
```
This shouldn't be necessary anymore after 5e923bc73da1322a2b64bfaf9fc090abfadfc57b.

anton-vlasenko replied:
I left it on purpose since I didn't know the code that validates the REST request parameters very well.
I've tested it a bit, and it looks like the validation is pretty reliable, and `$query_args['slug']` can't be a string.
Fixed in 005de4aa8ae617dc0df455d433861e7a83bf4bce

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -346,4 +342,40 @@ public function get_collection_params() {
+	 */
+	protected function get_transient_key( $query_args ) {
+
+		if ( array_key_exists( 'slug', $query_args ) ) {
```
It looks like this can be simplified to `isset`.

anton-vlasenko replied:
Since there was a test for
```php
$query_args['slug'] === null
```
in my unit tests, this made sense. But it's practically impossible to have a `null` value here. So I agree with you on this.
Fixed in 005de4aa8ae617dc0df455d433861e7a83bf4bce

anton-vlasenko committed:
Replace array_key_exist with isset because $query_args['slug'] isn't supposed to be null.

anton-vlasenko commented on src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -346,4 +342,39 @@ public function get_collection_params() {
+	protected function get_transient_key( $query_args ) {
+
+		if ( isset( $query_args['slug'] ) ) {
+			// This is an additional precaution because the "sort" function expects an array.
+			$query_args['slug'] = (array) $query_args['slug'];
```
I've added this typecasting because this method is protected. It means that any descendant class like [Gutenberg_REST_Pattern_Directory_Controller](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.0/class-gutenberg-rest-pattern-directory-controller.php) can change the REST schema. As a result, there may be a different value here (not necessarily an array).
If you think this is an overkill, please feel free to delete these 2 lines.

spacedmonkey replied:
```suggestion
			// This is an additional precaution because the "sort" function expects an array.
			$query_args['slug'] = wp_parse_list( $query_args['slug'] );
```

anton-vlasenko replied:
Fixed in 0a2c2ebfbfaf9de86be352e0957942f73b253f81

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -346,4 +342,39 @@ public function get_collection_params() {
+			sort( $query_args['slug'] );
+
+			// Slugs have to be imploded separately as implode doesn't work with recursive arrays.
+			$query_args['slug'] = implode( ',', $query_args['slug'] );
+
+			if ( '' === trim( $query_args['slug'] ) ) {
```
```suggestion
			$query_args['slug'] = trim( implode( ',', $query_args['slug'] ) );

			if ( '' === $query_args['slug'] ) {
```

anton-vlasenko replied:
In my opinion this code can be removed because of 2c08eeab216148827b02aa5f99a309948ca31a3b
`serialize` supports multidimensional arrays and we don't have to implode these values.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -346,4 +342,39 @@ public function get_collection_params() {
+			}
+		}
+
+		return 'wp_remote_block_patterns_' . md5( implode( '-', $query_args ) );
```
```suggestion
		return 'wp_remote_block_patterns_' . md5( serialize( $query_args ) );
```

anton-vlasenko replied:
Yes, I've been thinking about replacing `implode` with `wp_json_encode` (`json_encode` is faster than `serialize`) in the scope of another PR, but `serialize` will do too.
The result produced by `implode` doesn't depend on the keys of the `$query_args` array. Therefore, this increases the chance of collisions.
Fixed in 2c08eeab216148827b02aa5f99a309948ca31a3b
P.S. I didn't change it because I was afraid that it would cause a sharp increase in the number of requests to the `wp/v2/pattern-directory/patterns` endpoint once WordPress 6.0 is released. The old values stored in the cache will be invalidated because this PR replaces `implode` with `serialize`.
I hope it will not happen since not all WordPress sites are updated simultaneously.

spacedmonkey commented on tests/phpunit/tests/rest-api/rest-pattern-directory-controller.php:
```diff
@@ -399,6 +410,85 @@ public function test_get_item_schema() {
+	 * @param bool      $assert_same    Assertion type (assertSame vs assertNotSame).
+	 */
+	public function test_transient_keys_get_generated_correctly( $parameters_1, $parameters_2, $message, $assert_same = true ) {
+		$method = $this->get_reflection_method( self::$controller, 'get_transient_key' );
```
```suggestion
		$controller = new WP_REST_Pattern_Directory_Controller();
		$method = $this->get_reflection_method( $controller, 'get_transient_key' );
```

anton-vlasenko replied:
Refactored in 7d1c51221c42336c151aa454c7ce3b407b7e250f

spacedmonkey commented on tests/phpunit/tests/rest-api/rest-pattern-directory-controller.php:
```diff
@@ -21,6 +21,15 @@ class WP_REST_Pattern_Directory_Controller_Test extends WP_Test_REST_Controller_
 	 */
 	protected static $contributor_id;
 
+	/**
+	 * An instance of WP_REST_Pattern_Directory_Controller class.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @var WP_REST_Pattern_Directory_Controller
+	 */
+	private static $controller;
```
There is no need for this to be a class property here. It is only used in this test.

anton-vlasenko replied:
I agree, but I've refactored the test.
Now, this property is used in several test methods.

spacedmonkey commented on tests/phpunit/tests/rest-api/rest-pattern-directory-controller.php:
```diff
@@ -461,4 +551,21 @@ static function ( $return, $args, $url ) use ( $blocked_host ) {
+	 *
+	 * @return ReflectionMethod
+	 */
+	private function get_reflection_method( $object, $method ) {
```
Not sure if this needs to a function. It is only used in one place.

anton-vlasenko replied:
I agree. Fixed in 7d1c51221c42336c151aa454c7ce3b407b7e250f

spacedmonkey commented on tests/phpunit/tests/rest-api/rest-pattern-directory-controller.php:
```diff
@@ -461,4 +551,21 @@ static function ( $return, $args, $url ) use ( $blocked_host ) {
 		);
 	}
+
+	/**
+	 * Returns a reflection method and changes its scope.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param $object An object that has a private method that has to be called.
+	 * @param $method Name of the method.
+	 *
+	 * @return ReflectionMethod
+	 */
+	private function get_reflection_method( $object, $method ) {
```
Tests do not need these kind of comments.

anton-vlasenko replied:
Good to know.
Fixed in 7d1c51221c42336c151aa454c7ce3b407b7e250f

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -119,16 +119,7 @@ public function get_items( $request ) {
 			$query_args['slug'] = $slug;
```
Should this be passed through, `wp_parse_list`

anton-vlasenko replied:
Good point. Fixed in 0a2c2ebfbfaf9de86be352e0957942f73b253f81

anton-vlasenko replied:
It is passed through `wp_parse_list` here: https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/class-wp-rest-server.php#L983
More specifically, here: https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api.php#L1500
So, in theory, this is not needed.

spacedmonkey reviewed (changes requested):
There are a number of changes required here.

anton-vlasenko committed:
Pass $query_args['slugs'] through wp_parse_list to ensure that it is an array.

anton-vlasenko committed:
Refactor WP_REST_Pattern_Directory_Controller_Test.
Use WP_REST_Pattern_Directory_Controller_Test::setUp to set up the controller.

anton-vlasenko committed:
Refactor the test to use the static $controller property.
It seems there is some issue with the ::setUp method.

anton-vlasenko committed:
Use serialize instead of implode because transient keys must depend on request keys.

anton-vlasenko commented:
> There are a number of changes required here.

Thank you for the review, @spacedmonkey!
I've replied to all code review comments.
I'd appreciate the second round of the review.

spacedmonkey reviewed (approved):
LGTM

costdev commented on tests/phpunit/tests/rest-api/rest-pattern-directory-controller.php:
```diff
@@ -399,6 +408,106 @@ public function test_get_item_schema() {
+		$this->assertNotEmpty( $result_1 );
+
+		$this->assertIsString( $result_2 );
+		$this->assertNotEmpty( $result_2 );
```
Can we also add `message` parameters to these assertions in case of failure? It would immediately reveal which isn't a string, or which isn't empty.

anton-vlasenko replied:
Good point, @costdev.
Fixed in 5653f25aa2893fc715d5dd9be6ecd7483aa83e64

anton-vlasenko committed:
Improve WP_REST_Pattern_Directory_Controller_Test::test_transient_keys_get_generated_correctly: add error messages.

anton-vlasenko committed:
Improve WP_REST_Pattern_Directory_Controller_Test::test_transient_keys_get_generated_correctly: add error messages.

anton-vlasenko committed:
Update the test per @costdev 's suggestion.

anton-vlasenko committed:
Update the test per @costdev 's suggestion.

costdev reviewed (approved):
LGTM :+1:

ndiego commented:
@anton-vlasenko I'm working on cleaning up the [6.0 Project Board](https://github.com/orgs/WordPress/projects/27/views/1). Are we waiting on anything else for this PR?

anton-vlasenko commented:
> @anton-vlasenko I'm working on cleaning up the [6.0 Project Board](https://github.com/orgs/WordPress/projects/27/views/1). Are we waiting on anything else for this PR?

@ndiego 
From a technical point of view, this PR is ready to be merged.
2 core commiters have approved it.

I don't know if there are any problems from an organizational point of view.
I'd be grateful if you could please merge it (I don't know if you have the permission to merge to Beta 4).
Thank you.

cc @peterwilsoncc

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -346,4 +342,37 @@ public function get_collection_params() {
+			// This is an additional precaution because the "sort" function expects an array.
+			$query_args['slug'] = wp_parse_list( $query_args['slug'] );
+
+			// Sort the array so that the transient key doesn't depend on the order of slugs.
+			sort( $query_args['slug'] );
+
+			// Empty arrays should not affect the transient key.
+			if ( 0 === count( $query_args['slug'] ) ) {
+				unset( $query_args['slug'] );
+			}
```
Micro-optimization suggestions:

1. Only invoke `sort()` when the `array` has elements to sort.
2. Use `empty()` to detect that the array is empty. Why? (a) avoids the overhead with `count()`, (b) teeny tiny faster, (c) a wee bit more understandable, and (d) checks for a `false` return from `wp_parse_list()` (though highly unlikely a `false` will be returned.

```suggestion
			// Empty arrays should not affect the transient key.
			if ( empty( $query_args['slug'] ) ) {
				unset( $query_args['slug'] );
			} else {
				// Sort the array so that the transient key doesn't depend on the order of slugs.
				sort( $query_args['slug'] );
			}
```

See it in action https://3v4l.org/kOqb5.

hellofromtonya replied:
Resolved in https://github.com/WordPress/wordpress-develop/pull/2625/commits/1d8f893540d5fb7755444376d606059cfd677508.

anton-vlasenko replied:
I do agree with your suggestions, @hellofromtonya.
Thank you!
I will allow myself to comment on them :) 
1. I hoped that the "sort" function would have a check not to sort empty arrays. But I'm not 100% sure. So, yes, let's not call it on empty arrays.
2. Well, this is news to me. I thought that the `count` function is faster because it knows the type of the input data( i.e., arrays), unlike the `empty` function, which does not initially know the type of the input data. I was wrong. That's good to know.

anton-vlasenko committed:
Micro-optimisations per @hellofromtonya 's suggestion here: https://github.com/WordPress/wordpress-develop/pull/2625#discussion_r862808359

hellofromtonya reviewed (approved):
Ready for commit.

hellofromtonya commented:
Committed via changeset https://core.trac.wordpress.org/changeset/53333.

demahes123 commented:
I'm working on cleaning up the [6.0 Project Board](https://github.com/orgs/WordPress/projects/27/views/1). Are we waiting on anything else for this PR?
{{{
#!html
<a href="https://trademinor.org/"></a><a href="https://bitcoinistltd.com/"></a><a href="https://totimescanews.ca"></a><a href="https://hnmag.co.uk/"></a><a href="https://ukhorizon.co.uk"></a><a href="https://britanniabloom.co.uk/"></a><a href="https://britishvista.co.uk/"></a><a href="https://englandecho.co.uk/"></a><a href="https://viralreel.org/"></a><a href="https://fool.social"></a><a href="https://jobsinfoall.com/"></a><a href="http://myblogvista.com/"></a><a href="https://mytimesworld.com"></a><a href="https://networkustad.co.uk"></a><a href="https://jkjobalerts.com/"></a><a href="https://sgvapeoutlet.com/"></a><a href="https://itsdailytimes.com/"></a><a href="https://saabry.com/"></a><a href="https://gbcspa.com"></a>
}}}
